ConditionalCompilation.md

```diff
@@ -1,28 +1,140 @@
+
+ * **unstructured**: arbitrary code sections can be injected or conditionaly included. This is what the C preprocessor and macros do, and a lesser version of that is proposed in [`simple templating`](SimpleTemplating.md). 
+ * **structured**: only structural elements of the syntax can be manipulated, e.g. a whole declaration, a member, etc. Rust uses this approach with the `#[cfg]` attribute.
+
+We think that *structured* is the way to go. It leads to clearer and safer code, which is more important than implementation complexity in our eyes.
+Also, a language that has a good expressive power already should not need a way to hack around with arbitrary code injection, like C macros do.
+
+| unstructured | structured |
+|--------------|------------|
+| (+) much easier to implement: just look for the `#word` symbols | (-) harder to implement: requires a full parsing |
+| (+) often language-agnostic (see the [C preprocessor](https://en.wikipedia.org/wiki/C_preprocessor). Familiar to C developers. | (-) a new syntax needs to be taught, not always self-explanatory |
+| (-) poorly integrated in the language, harder to read by humans | (+) is a well-designed syntax feature of the language |
+| (+) technically more expressive (e.g. [manipulating identifiers](https://en.wikipedia.org/wiki/C_preprocessor#Token_concatenation)) | (-) can only conditionaly include parts of the syntax tree, poor text generation capabilities. |
+| (-) behaves unpredictably, intent and behavior is hidden | (+) intent and behavior is visible at the usage site |
+| (-) no type checking, no syntax checking, because code is generated dynamically | (+) can statically check that conditional variants lead to syntactically correct and type-valid code |
+| (-) poor IDE support | (+) IDE can check all possible code paths |
+
+## Proposal
+
+We propose to leverage the [*attribute* syntax](https://www.w3.org/TR/WGSL/#attributes) to decorate syntax nodes with conditional compilation attributes.
+The compiler is provided with a set of '*compile-time features*' to enable (see below).
+
+### Definitions
+
+ * **Compile-time expression**: A *compile-time expression* is evaluated by the compiler before the rest of the program. It lives in the *compile-time scope*.
+
+ * **Compile-time scope**: The *compile-time scope* this is an independent scope from the [*module scope*](https://www.w3.org/TR/WGSL/#module-scope), meaning it cannot see any declarations from the source code, and its identifers are independent.
+
+ * **Compile-time feature**: A *compile-time feature* is an identifier that evaluates to a boolean. It is set to `true` if the feature is *enabled* during the compilation phase.
+
+ * **Compile-time attribute**: A *compile-time attribute* is parametrized by a *compile-time expression*. It is eliminated after the compilation phase but can affect the syntax node it decorates.
```

Comment:
How do we handle runtime conditions?

Reply:
Discord discussion: 
```
say we invoke the compiler/linker this way:

const features = { feat1: true, feat2: false, feat3: undefined }
const compiled = linker.link(source, features)

then this file:

@if(feat1)
fn func1() {}

@if(feat2)
fn func2() {}

@if(feat1 && feat3)
@if func3() {}

@if(feat2 && feat3)
@if func4() {}


would be compiled like this:

fn func1() {}

// func2 is eliminated

@if(true && feat3) // attribute partially resolved
@if func3() {}

@if(false && feat3) // attribute partially resolved
@if func4() {}
```

I'll add this to the spec in the execution section.

Reply:
"runtime conditions" as in "we run the compiler at runtime, instead of having an offline compilation pass"?

And not "runtime" as in "runtime of the shader", right?

Reply:
I understood the former, i.e. for example a feature might depend on a runtime condition (e.g. supported hardware). In this case one has two choices: generate all possible shader variants or compile at runtime (the start of the program).

We could do a mix of that, generating a partially compiled wesl shader (all imports resolved, most of the cond-comp resolved) and run a cheap 2nd compilation with the missing featuers.

Reply:
Exactly - I meant running the compiler at runtime vs build time.

While partially compiling at build time and then completing the compile at runtime is possible, I'd suggest we skip mixed mode until we need it.